### PR TITLE
feat: Docker target platform override

### PR DESCRIPTION
**Pull request by shivani170**
# Description
Docker target platform override 

Fixes [AB#1643](https://dev.azure.com/DevtronLabs/Devtron/_boards/board/t/Devtron%20Team/Stories/?workitem=1643)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
- Create a new ci pipeline in the workflow 
- Click on allow override
- Override the target platform(with customize platform as well)
- then update and check the values

# Checklist:

* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/CIPipelineN/AdvancedConfigOptions.tsx [hunk 1]
```diff
@@ -179,6 +181,51 @@ export default function AdvancedConfigOptions({
...
+    const [showCustomPlatformWarning, setShowCustomPlatformWarning] = useState<boolean>(false)
+
+    const updateSelectedPlatformsCustomTargetPlatform = () => {
+        let _customTargetPlatform = false
+        let _selectedPlatforms = []
+        let targetPlatform = ''
+        if (parentState.selectedCIPipeline?.dockerConfigOverride?.ciBuildConfig?.dockerBuildConfig?.targetPlatform) {
+            targetPlatform =
+                parentState.selectedCIPipeline.dockerConfigOverride.ciBuildConfig.dockerBuildConfig.targetPlatform
+        } else if (parentState.ciConfig?.ciBuildConfig) {
```
Comment (vivek-devtron) [inline src/components/CIPipelineN/AdvancedConfigOptions.tsx:195]: we can combine these 2 conditions

File: src/components/CIPipelineN/AdvancedConfigOptions.tsx [hunk 2]
```diff
@@ -179,6 +181,51 @@ export default function AdvancedConfigOptions({
...
+            platformsArray.push(o.label)
+        })
+        if (_form.dockerConfigOverride?.ciBuildConfig?.dockerBuildConfig) {
+            _form.dockerConfigOverride.ciBuildConfig.dockerBuildConfig.targetPlatform = platformsArray.join()
+        }
+        setFormData(_form)
+    }
+
+    useEffect(() => {
+        updateSelectedPlatformsCustomTargetPlatform()
```
Comment (vivek-devtron) [inline src/components/CIPipelineN/AdvancedConfigOptions.tsx:221]: we should add a failsafe empty check on parentState

File: src/components/CIPipelineN/AdvancedConfigOptions.tsx [hunk 3]
```diff
@@ -179,6 +181,51 @@ export default function AdvancedConfigOptions({
...
+    }, [parentState])
+
+    useEffect(() => {
+        handleChangeInTargetPlatforms()
```
Comment (vivek-devtron) [inline src/components/CIPipelineN/AdvancedConfigOptions.tsx:225]: we should add a failsafe empty check on selectedTargetPlatforms

File: src/components/ciConfig/TargetPlatformSelector.tsx [hunk 4]
```diff
@@ -0,0 +1,162 @@
+import React from 'react'
+import { components } from 'react-select'
+import CreatableSelect from 'react-select/creatable'
+import { ReactComponent as WarningIcon } from '../../assets/icons/ic-warning.svg'
+import { TARGET_PLATFORM_LIST, tempMultiSelectStyles } from './CIConfig.utils'
+
+export interface TargetPlatformSelector {
```
Comment (vivek-devtron) [inline src/components/ciConfig/TargetPlatformSelector.tsx:7]: move this to types

File: src/components/ciConfig/TargetPlatformSelector.tsx [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+function TargetPlatformSelector({
+    allowOverride,
+    selectedTargetPlatforms,
+    setSelectedTargetPlatforms,
+    showCustomPlatformWarning,
+    setShowCustomPlatformWarning,
+    targetPlatformMap,
+    parentState,
+}: TargetPlatformSelector) {
+    const platformMenuList = (props): JSX.Element => {
```
Comment (vivek-devtron) [inline src/components/ciConfig/TargetPlatformSelector.tsx:26]: extract `platformMenuList` out from this component as sometime this cause re renders

File: src/components/ciConfig/TargetPlatformSelector.tsx [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        return (
+            <components.MenuList {...props}>
+                <div className="cn-5 pl-12 pt-4 pb-4 dc__italic-font-style">
+                    Type to enter a target platform. Press Enter to accept.
```
Comment (vivek-devtron) [inline src/components/ciConfig/TargetPlatformSelector.tsx:30]: move messaging to const

File: src/components/ciConfig/TargetPlatformSelector.tsx [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        // let targetPlatform =
+        //     parentState?.selectedCIPipeline?.dockerConfigOverride?.ciBuildConfig?.dockerBuildConfig?.targetPlatform
+        let targetPlatform =
+            parentState?.ciConfig?.ciBuildConfig?.dockerBuildConfig?.targetPlatform
+            if(!targetPlatform){
+             return <div className='bcn-1 br-4 flex cn-7 pt-8 pb-8'>
+                Target platform is not set
+              </div>
+            }else{
+              if (targetPlatform && !targetPlatform?.includes(',')) {
```
Comment (vivek-devtron) [inline src/components/ciConfig/TargetPlatformSelector.tsx:101]: no need to add check for `targetPlatform` as it is already getting checked in if

File: src/components/CIPipelineN/AdvancedConfigOptions.tsx (hunk 1 above)
Comment (gireesh-devtron) [reply src/components/CIPipelineN/AdvancedConfigOptions.tsx]: Done

File: src/components/CIPipelineN/AdvancedConfigOptions.tsx (hunk 2 above)
Comment (gireesh-devtron) [reply src/components/CIPipelineN/AdvancedConfigOptions.tsx:221]: added the failsafe check

File: src/components/CIPipelineN/AdvancedConfigOptions.tsx (hunk 3 above)
Comment (gireesh-devtron) [reply src/components/CIPipelineN/AdvancedConfigOptions.tsx:225]: added the failsafe check

File: src/components/ciConfig/TargetPlatformSelector.tsx (hunk 5 above)
Comment (shivani170) [reply src/components/ciConfig/TargetPlatformSelector.tsx:28]: doen

File: src/components/ciConfig/TargetPlatformSelector.tsx (hunk 6 above)
Comment (shivani170) [reply src/components/ciConfig/TargetPlatformSelector.tsx:32]: Done

File: src/components/ciConfig/TargetPlatformSelector.tsx (hunk 7 above)
Comment (shivani170) [reply src/components/ciConfig/TargetPlatformSelector.tsx:101]: done

File: src/components/ciConfig/TargetPlatformSelector.tsx [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+    const getOverridenValue = () => {
+        if (!targetPlatform) {
+            return (
+                <div className="bcn-1 br-4 flex cn-7 pt-8 pb-8">
+                    {SelectorMessaging.PALTFORM_DESCRIPTION_WITH_NO_TARGET}
+                </div>
+            )
+        } else {
+            if (!targetPlatform?.includes(',')) {
```
Comment (vivek-devtron) [inline src/components/ciConfig/TargetPlatformSelector.tsx:99]: No need to add optional check on targetPlatform.

File: src/components/ciConfig/TargetPlatformSelector.tsx [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+                return <div className="en-2 bw-1 br-4 dc__w-fit-content pl-8 pr-8 pt-2 pb-2 mr-8">{targetPlatform}</div>
+            } else {
+                return (
+                    <div className="flex left ">
+                        {targetPlatform?.split(',').map((val) => {
```
Comment (vivek-devtron) [inline src/components/ciConfig/TargetPlatformSelector.tsx:104]: Same here

File: src/components/ciConfig/TargetPlatformSelector.tsx [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    selectedTargetPlatforms,
+    setSelectedTargetPlatforms,
+    showCustomPlatformWarning,
+    setShowCustomPlatformWarning,
+    targetPlatformMap,
+    targetPlatform,
+    configOverrideView,
+    updateDockerConfigOverride,
+}: TargetPlatformSelectorType) {
+    const noMatchingPlatformOptions = (): string => {
```
Comment (vivek-devtron) [inline src/components/ciConfig/TargetPlatformSelector.tsx:30]: We should move it in helper file as this has been used on multiple places

Comment (shivani170) [reply src/components/ciConfig/TargetPlatformSelector.tsx:30]: Done

**Pull request merged**